apps/src/code-studio/components/progress/teacherPanel/TeacherPanel.jsx

```diff
@@ -0,0 +1,378 @@
+      ...overrideData
+    };
+
+    firehoseClient.putRecord({
+      study: 'teacher_panel',
+      event: eventName,
+      data_json: JSON.stringify(data)
+    });
+  };
+
+  onSelectUser = (id, selectType) => {
+    this.logToFirehose('select_student', {select_type: selectType});
+    this.props.onSelectUser(id);
+  };
+
+  getSectionData = onComplete => {
+    const {
+      sectionData,
+      levelId,
+      lessonId,
+      scriptId,
+      isLessonExtras
+    } = this.props;
+
+    if (!sectionData?.section) {
+      return;
+    }
+
+    this.setState({isLoadingLevelsWithProgress: true});
+
+    const baseUrl = `/api/teacher_panel_progress/${sectionData.section.id}`;
+
+    let query;
+    if (isLessonExtras) {
+      query = {
+        lesson_id: lessonId,
+        is_lesson_extras: true,
+        script_id: scriptId
+      };
+    } else {
+      query = {
+        script_id: scriptId,
+        level_id: levelId
+      };
+    }
+
+    $.ajax({
+      url: baseUrl + '?' + queryString.stringify(query),
+      method: 'GET'
+    })
+      .done(data => {
+        this.setState({
+          levelsWithProgress: data,
+          isLoadingLevelsWithProgress: false
+        });
+        onComplete && onComplete();
+      })
+      .fail(err => {
+        console.log(
+          `Failed to update teacher panel (${err.status}) ${err.statusText}`
+        );
+        this.setState({isLoadingLevelsWithProgress: false});
+      });
+  };
+
+  render() {
+    const {
+      sectionData,
+      viewAs,
+      hasSections,
+      sectionsAreLoaded,
+      selectedSection,
+      unitHasLockableLessons,
+      unlockedLessonNames,
+      students,
+      unitName
+    } = this.props;
+
+    let currentStudent = null;
+    let currentStudentScriptLevel = null;
+
+    const {levelsWithProgress} = this.state;
+
+    if (sectionData) {
+      if (sectionData.section && sectionData.section.students) {
+        currentStudent = sectionData.section.students.find(
+          student => this.props.getSelectedUserId() === student.id
+        );
+
+        if (currentStudent) {
+          if (levelsWithProgress) {
+            currentStudentScriptLevel = levelsWithProgress.find(
+              level => this.props.getSelectedUserId() === level.userId
+            );
+          }
+        } else {
+          currentStudent = {
+            id: null,
+            name: i18n.studentTableTeacherDemo()
+          };
+          currentStudentScriptLevel = sectionData.teacher_level;
+        }
+      }
+    }
+
+    const sectionId = selectedSection && selectedSection.id;
+
+    const displaySelectedStudentInfo =
+      viewAs === ViewType.Teacher &&
+      currentStudent &&
+      (students || []).length > 0;
+
+    const displayLevelExamples =
+      viewAs === ViewType.Teacher && sectionData && sectionData.level_examples;
+
+    const displayLockInfo =
+      hasSections && unitHasLockableLessons && viewAs === ViewType.Teacher;
```

Comment:
(very minor) Maybe move the viewAs condition to be the first in the list to be more consistent with the other two checks above?  That should make it clear that all of these only apply to `ViewType.Teacher`.

We probably don't want to fix this now, but I'm also wondering how sectionData relates to hasSections (i.e. can one be derived from the other)?

Reply:
I created this ticket https://codedotorg.atlassian.net/browse/LP-1987, there are still many aspects of the teacher panel logic and how data is included that are confusing, hopefully I'll get a chance to address some of this there.